src/ServiceBusExplorer/Helpers/ServiceBusHelper.cs

```
@@ -1872,14 +1872,38 @@ public IEnumerable<QueueDescription> GetQueues(string filter)
-                    if (task.IsCompleted)
+                    //Documentation states AND is the only logical clause allowed in the filter (And FYI a maximum of only 3 filter expressions allowed)
+                    //https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.namespacemanager.getqueuesasync?view=azure-dotnet#Microsoft_ServiceBus_NamespaceManager_GetQueuesAsync_System_String_
+                    //Split on ' OR ' and combine queues returned
+                    IEnumerable<QueueDescription> queues = new List<QueueDescription>();
+                    var tasksCompleted = true;
+                    var filters = new List<string>();
+                    if (string.IsNullOrWhiteSpace(filter))
                     {
-                        return task.Result;
+                        filters.Add(filter);
+                    }
+                    else
+                    {
+                        filters = filter.ToLowerInvariant().Split(new[] { " or " }, StringSplitOptions.None).ToList();
+                    }
+                    foreach (var splitFilter in filters)
+                    {
+                        var task = string.IsNullOrWhiteSpace(filter) ? namespaceManager.GetQueuesAsync() : namespaceManager.GetQueuesAsync(splitFilter);
+                        taskList.Add(task);
+                        taskList.Add(Task.Delay(TimeSpan.FromSeconds(MainForm.SingletonMainForm.ServerTimeout)));
+                        Task.WaitAny(taskList.ToArray());
+                        if (task.IsCompleted)
+                        {
+                            queues = queues.Union(task.Result);
+                            taskList.Clear();
+                        }
+                        else
+                        {
+                            tasksCompleted = false;
+                            break;
+                        }
                     }
+                    if (tasksCompleted) return queues;
```

Is `tasksCompleted` is necessary given the needed information is found with `task.IsCompleted`?

---

Hi Sean, 

Yes, as there are potentially multiple iterations reusing the 'task' variable, any one of which could timeout which we would want to result in the whole operation raising a TimeoutException and not return a partially complete list of queues or topics. 

It's there to ensure all calls to namespaceManager.GetQueuesAsync(splitFilter) are successful and to fail fast if any of them fail.

---

I see what you mean now. Refactored tasksCompleted out of queues and topics code now